File: python/src/prts_mcp/data/story_memoir.py
"""Operator memoir discovery via chardict.json reverse lookup.

Split from story.py. Resolves an operator Chinese name to its internal
code, then finds all memoir events (story_{code}_set_N) and their chapters.
"""
from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path

from prts_mcp.data.stores import DirectoryStore, JsonStore, ZipStore
from prts_mcp.data.story_reader import (
    CHARDICT,
    STORY_REVIEW_TABLE,
    MemoirChapter,
    OperatorMemoirResult,
    load_json,
    story_store,
)

# ---------------------------------------------------------------------------
# Chardict loading + caching
# ---------------------------------------------------------------------------


def _load_chardict(store: JsonStore) -> dict[str, dict[str, object]]:
    """Load chardict.json from the story store."""
    descriptor = _chardict_store_descriptor(store)
    if descriptor is not None:
        return _cached_chardict(descriptor)
    if not store.exists(CHARDICT):
        return {}
    return _read_chardict_from_store(store)


def _read_chardict_from_store(store: JsonStore) -> dict[str, dict[str, object]]:
    try:
        data: dict = load_json(store, CHARDICT)  # type: ignore[assignment]
        return {str(k): v for k, v in data.items() if isinstance(v, dict)}
    except (KeyError, FileNotFoundError, json.JSONDecodeError):
        return {}


def _chardict_store_descriptor(store: JsonStore) -> tuple[str, str, int, int] | None:
    if isinstance(store, ZipStore):
        path = store.zip_path.resolve()
        if not path.is_file():
            return None
        stat = path.stat()
        return ("zip", str(path), stat.st_size, stat.st_mtime_ns)
    if isinstance(store, DirectoryStore):
        root = store.root.resolve()
        chardict = root / CHARDICT
        if not chardict.is_file():
            return None
        stat = chardict.stat()
        return ("directory", str(root), stat.st_size, stat.st_mtime_ns)
    return None


@lru_cache(maxsize=4)
def _cached_chardict(
    descriptor: tuple[str, str, int, int],
) -> dict[str, dict[str, object]]:
    kind, source, _size, _mtime_ns = descriptor
    store: JsonStore
    if kind == "zip":
        store = ZipStore(source)
    else:
        store = DirectoryStore(source)
    try:
        return _read_chardict_from_store(store)
    finally:
        store.close()


def clear_chardict_cache() -> None:
    """Clear the cached chardict."""
    _cached_chardict.cache_clear()


# ---------------------------------------------------------------------------
# Code resolution
# ---------------------------------------------------------------------------


def _resolve_operator_code(
    chardict: dict[str, dict[str, object]],
    operator_name: str,
) -> str | None:
    """Reverse lookup: operator display name -> internal code (e.g. '能天使' -> 'angel')."""
    for code, info in chardict.items():
        if info.get("name") == operator_name:
            return code
    return None


# ---------------------------------------------------------------------------
# Public API — zip-path wrapper
# ---------------------------------------------------------------------------


def get_operator_memoirs(
    zip_path: Path,
    operator_name: str,
) -> OperatorMemoirResult:
    """Return all memoir chapters for an operator by Chinese name.

    Raises:
        KeyError: If the operator is not found in chardict or has no memoirs.
    """
    with story_store(zip_path) as store:
        return get_operator_memoirs_from_store(store, operator_name)


# ---------------------------------------------------------------------------
# Public API — store-based
# ---------------------------------------------------------------------------


def get_operator_memoirs_from_store(
    store: JsonStore,
    operator_name: str,
) -> OperatorMemoirResult:
    """Return all memoir chapters for an operator by Chinese name using a JSON store."""
    chardict = _load_chardict(store)
    if not chardict:
        raise KeyError("chardict.json 未在 story zip 中找到。")

    code = _resolve_operator_code(chardict, operator_name)
    if code is None:
        raise KeyError(
            f"未找到干员名称 '{operator_name}' 对应的内部代码。请使用游戏内中文名称。"
        )

    char_info = chardict[code]
    operator_id = char_info.get("id", "")

    table: dict = load_json(store, STORY_REVIEW_TABLE)  # type: ignore[assignment]
    prefix = f"story_{code}_set_"
    memoir_events = sorted(
        (eid for eid in table if eid.startswith(prefix)),
        key=lambda eid: int(eid.rsplit("_", 1)[-1]) if eid.rsplit("_", 1)[-1].isdigit() else 0,
    )

    if not memoir_events:
        raise KeyError(f"干员 '{operator_name}' (code={code}) 暂无密录数据。")

    chapters: list[MemoirChapter] = []
    for event_id in memoir_events:
        entry = table[event_id]
        datas = sorted(
            entry.get("infoUnlockDatas") or [],
            key=lambda x: x.get("storySort", 0),
        )
        for d in datas:
            story_key = d.get("storyTxt")
            if not story_key:
                continue
            chapters.append(MemoirChapter(
                event_id=event_id,
                story_key=story_key,
                story_code=d.get("storyCode", ""),
                story_name=d.get("storyName", ""),
            ))

    return OperatorMemoirResult(
        operator_name=operator_name,
        internal_code=code,
        operator_id=operator_id,
        total_chapters=len(chapters),
        chapters=chapters,
    )

File: ts/src/data/storyMemoir.ts
/**
 * Operator memoir discovery via chardict.json reverse lookup.
 *
 * Split from story.ts. Resolves an operator Chinese name to its internal
 * code, then finds all memoir events (story_{code}_set_N) and their chapters.
 * Mirrors python/src/prts_mcp/data/story_memoir.py.
 */

import { statSync } from "node:fs";
import { join } from "node:path";
import { DirectoryStore, type JsonStore, ZipStore } from "./stores.js";
import {
  CHARDICT,
  type MemoirChapter,
  type OperatorMemoirResult,
  type RawReviewTable,
  STORY_REVIEW_TABLE,
  storyStore,
  withStoryStore,
} from "./storyReader.js";

export interface OperatorMemoirsPayload {
  operator_name: string;
  internal_code: string;
  operator_id: string;
  total: number;
  chapters: Array<{
    story_code: string;
    story_name: string;
    story_key: string;
  }>;
}

// ---------------------------------------------------------------------------
// Chardict types + cache
// ---------------------------------------------------------------------------

interface CharDictEntry {
  name?: string;
  id?: string;
}

let charDictCache: { descriptor: string; data: Record<string, CharDictEntry> } | null = null;

export function clearCharDictCache(): void {
  charDictCache = null;
}

// ---------------------------------------------------------------------------
// Chardict loading
// ---------------------------------------------------------------------------

function loadCharDict(store: JsonStore): Record<string, CharDictEntry> {
  const descriptor = charDictStoreDescriptor(store);
  if (descriptor !== null && charDictCache?.descriptor === descriptor) {
    return charDictCache.data;
  }
  if (!store.exists(CHARDICT)) {
    return {};
  }
  let data: Record<string, CharDictEntry>;
  try {
    data = store.readJson<Record<string, CharDictEntry>>(CHARDICT);
  } catch {
    data = {};
  }
  if (descriptor !== null) charDictCache = { descriptor, data };
  return data;
}

function charDictStoreDescriptor(store: JsonStore): string | null {
  if (store instanceof ZipStore) {
    try {
      const stat = statSync(store.zipPath, { bigint: true });
      return `zip:${store.zipPath}:${stat.size}:${stat.mtimeNs}:chardict`;
    } catch {
      return null;
    }
  }
  if (store instanceof DirectoryStore) {
    const chardict = join(store.root, CHARDICT);
    try {
      const stat = statSync(chardict, { bigint: true });
      return `directory:${store.root}:${stat.size}:${stat.mtimeNs}:chardict`;
    } catch {
      return null;
    }
  }
  return null;
}

// ---------------------------------------------------------------------------
// Code resolution
// ---------------------------------------------------------------------------

function resolveOperatorCode(
  charDict: Record<string, CharDictEntry>,
  operatorName: string
): string | null {
  for (const [code, info] of Object.entries(charDict)) {
    if (info.name === operatorName) return code;
  }
  return null;
}

// ---------------------------------------------------------------------------
// Public API — zip-path wrapper
// ---------------------------------------------------------------------------

export function getOperatorMemoirs(
  zipPath: string,
  operatorName: string,
): OperatorMemoirResult {
  return withStoryStore(zipPath, (store) => getOperatorMemoirsFromStore(store, operatorName));
}

export function buildOperatorMemoirs(
  zipPath: string,
  operatorName: string,
): OperatorMemoirsPayload {
  return withStoryStore(zipPath, (store) => buildOperatorMemoirsFromStore(store, operatorName));
}

export function renderOperatorMemoirs(data: OperatorMemoirsPayload): string {
  const lines: string[] = [
    `# ${data.operator_name}（code: ${data.internal_code}，id: ${data.operator_id}）`,
    `共 ${data.total} 章密录\n`,
  ];
  for (const ch of data.chapters) {
    lines.push(`- ${ch.story_code} ${ch.story_name}（key: ${ch.story_key}）`);
  }
  return lines.join("\n");
}

// ---------------------------------------------------------------------------
// Public API — store-based
// ---------------------------------------------------------------------------

export function getOperatorMemoirsFromStore(
  store: JsonStore,
  operatorName: string,
): OperatorMemoirResult {
  const charDict = loadCharDict(store);
  if (Object.keys(charDict).length === 0) {
    throw new Error("chardict.json 未在 story zip 中找到。");
  }

  const code = resolveOperatorCode(charDict, operatorName);
  if (code === null) {
    throw new Error(
      `未找到干员名称 '${operatorName}' 对应的内部代码。请使用游戏内中文名称。`
    );
  }

  const charInfo = charDict[code];
  const operatorId = charInfo.id ?? "";

  const table = store.readJson<RawReviewTable>(STORY_REVIEW_TABLE);
  const prefix = `story_${code}_set_`;
  const memoirEvents = Object.keys(table)
    .filter((eid) => eid.startsWith(prefix))
    .sort((a, b) => {
      const aNum = parseInt(a.split("_").pop() ?? "0", 10) || 0;
      const bNum = parseInt(b.split("_").pop() ?? "0", 10) || 0;
      return aNum - bNum;
    });

  if (memoirEvents.length === 0) {
    throw new Error(`干员 '${operatorName}' (code=${code}) 暂无密录数据。`);
  }

  const chapters: MemoirChapter[] = [];
  for (const eventId of memoirEvents) {
    const entry = table[eventId];
    const datas = (entry?.infoUnlockDatas ?? []).slice();
    datas.sort((a, b) => (a.storySort ?? 0) - (b.storySort ?? 0));
    for (const d of datas) {
      if (!d.storyTxt) continue;
      chapters.push({
        eventId,
        storyKey: d.storyTxt,
        storyCode: d.storyCode ?? "",
        storyName: d.storyName ?? "",
      });
    }
  }

  return {
    operatorName,
    internalCode: code,
    operatorId,
    totalChapters: chapters.length,
    chapters,
  };
}

export function buildOperatorMemoirsFromStore(
  store: JsonStore,
  operatorName: string,
): OperatorMemoirsPayload {
  const result = getOperatorMemoirsFromStore(store, operatorName);
  return {
    operator_name: result.operatorName,
    internal_code: result.internalCode,
    operator_id: result.operatorId,
    total: result.totalChapters,
    chapters: result.chapters.map((ch) => ({
      story_code: ch.storyCode,
      story_name: ch.storyName,
      story_key: ch.storyKey,
    })),
  };
}
